feat: SocialView 구현

parkyounbae opened the pull request:
유저나 친구의 데이터를 표시하는 테이블 뷰
잔디를 표시하는 컬렌션 뷰
친구를 추가하는 버튼 임시 구현

Resolve: GITIT-52

parkyounbae committed:
feat: SocialView 구현

유저나 친구의 데이터를 표시하는 테이블 뷰
잔디를 표시하는 컬렌션 뷰
친구를 추가하는 버튼 임시 구현

Resolve: GITIT-52

parkyounbae committed:
fix: 버튼의 크기가 안맞고 움직이는 문제 수정

fontSize를 수정
오토레이아웃 수정

Resolve: GITIT-52

parkyounbae committed:
feat: 날짜에 따른 잔디 표현 기능 구현, 친구의 잔디 표현 기능 구현

임시 데이터로 테스트
레이아웃 수정 필요

Resolve: GITIT-52

parkyounbae committed:
feat: 잔디가 표현하는 기간을 1년으로 늘림

잔디의 갯수를 추가하고 스크롤을 끝으로 가게 해놓음
잔대 셀이 재사용 될때 색이 칠해진 셀이 그대로 재사용 되는것을 막음

Resolve: GITIT-52

parkyounbae committed:
feat: 친구 테이블 삭제 기능 임시 구현, 버튼 알람 기능 추가, 프로퍼티 변경

본인을 제외한 친구 셀을 스와이프 하여 삭제하는 기능 임시 구현
친구 5명 이상일때 버튼을 누르면 Alert 뜨는 기능 구현
SocialViewFriendsTableViewCell 에서 친구와 유저의 정보를 따로받지 않고 통일된 프로퍼티로 받음
UI 일부 수정

Resolves: GITIT-52

parkyounbae committed:
feat: 서버에서 친구리스트를 받아오고난 뒤 커밋 오름차순으로 정렬

오름차순으로 정렬

Resolves: GITIT-52

parkyounbae committed:
style: 불필요한 프로퍼티 삭제

friendlist를 별도의 프로퍼티를 생성하지 않고 userinfo의 것을 가져옴

Resolves: GITIT-52

sweethoneybee commented on Git-it/ViewController/Cell/SocialViewFriendsTableViewCell.swift:
```diff
@@ -0,0 +1,156 @@
+}
+
+// MARK: - Extension
+
```
View가 VC역할도 맡아서, MVC 패턴에 위배중인데,
우선은 이 구현 유지하는 걸로 해요 인지하고 계시다가 나중에 수정하는 걸로

sweethoneybee commented on Git-it/ViewController/Cell/SocialViewFriendsTableViewCell.swift:
```diff
@@ -0,0 +1,156 @@
+    static let identifier = "SocialViewFriendsTableViewCell"
+    var userNameLabel: UILabel?
+    var grassCollectionView: GrassCollectionView?
+    var currentDateIndex: Int {
```
프로퍼티 중에 하나의 메소드에서만 쓰이는 프로퍼티(dateFormatter)는
해당 메소드의 지역변수로 빼는 게 좋을 듯해요

parkyounbae replied:
수정 완료했습니다.

sweethoneybee commented on Git-it/ViewController/Cell/SocialViewFriendsTableViewCell.swift:
```diff
@@ -0,0 +1,156 @@
+
+        self.grassCollectionView?.collectionViewLayout = flowLayout
+        
+        DispatchQueue.main.async {
```
collectionViewCellFlowLayout 메소드는 메인큐 외의 흐름에서 실행될 여지가 없어서
굳이 이 부분을 메인큐에 넣어주지 않아도 괜찮을 듯해요
결과물의 차이가 사실상 없겠지만, 메인큐에서 비동기로 처리하고 있는 이유가 있는지 궁금해요

parkyounbae replied:
이 부분은 저도 잘 모르긴 합니다. 아무리 scrollToItem을 해도 적용이 안되길래 구글링좀 해봤는데 여러가지 답변중에서 비동기식으로 적용하는것만 저한테는 통하더라구요 ㅋㅋㅋ 정작 그 댓글에서도 이유는 모른다고...

sweethoneybee replied:
아하 찾아보니 ios14 버전에서 버그가 있다는 이야기도 있네요. 다른 여러 방식들 이미 시도해보셨을 테니 ㅋㅋㅋㅋ.
지금 이 방식이 잘 되어서 다행이네요

sweethoneybee commented on Git-it/ViewController/Cell/SocialViewFriendsTableViewCell.swift:
```diff
@@ -0,0 +1,156 @@
+    
+    // MARK: - Init
+
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
```
여기서 cellForRowAt 메소드에서 호출하던 메소드 호출해주는 게 더 깔끔할 듯해요(뷰와 관련된 메소드들)

parkyounbae replied:
수정 했습니다~

sweethoneybee commented on Git-it/ViewController/Cell/SocialViewFriendsTableViewCell.swift:
```diff
@@ -0,0 +1,156 @@
+    
+    // MARK: - Init
+
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
```
init은 메소드의 젤 위쪽으로 올려주는 게 좋을 듯해요. init, 상속메소드만큼은 위에.. ㅋㅋㅋ

parkyounbae replied:
수정했습니다~

sweethoneybee commented on Git-it/ViewController/Cell/SocialViewFriendsTableViewCell.swift:
```diff
@@ -0,0 +1,156 @@
+        }
+        
+        if let records = userCommitRecords {
+            for userdata in records {
```
**실제 데이터를 받는다고 생각하고 구현을 수정할 필요가 있을 것 같아요**

이 메소드에서는 무거운 작업을 하면 안돼요. 빠르게 셀을 만들고 리턴해줘야 해요.  
현재는 샘플데이터가 커밋한 날에 해당하는 데이터만 담고 있지만
**실제 데이터는 커밋하지 않은 날도 모두 데이터로 담아서(아마 `count`, `level`을 0으로 주겠죠?),  
콜렉션뷰의 아이템수와 동일하게 줄 거예요.** 그러면 지금처럼 `userCommitsRecords` 데이터를 
순회할 필요가 없어지고 `indexPath.item`을 인덱스 값으로 해서 필요한 데이터에 접근할 수 있을 거예요

sweethoneybee commented on Git-it/ViewController/Cell/GrassCollectionViewCell.swift:
```diff
@@ -0,0 +1,42 @@
+    static let identifier = "grassCollectionViewCell"
+    var commitLevel: Int?
+    
+    func setColor() {
```
**뷰와 데이터는 최대한 분리해주세요**

저도 셀에 데이터 저장해두고, 그 값에 따라 셀이 색 변경되도록 구현하곤 했는데
뷰와 데이터는 분리되어 있는 게 더 MVC 패턴에 맞는 것 같아요
그래서 commitLetvel을 뷰에서 들고 있기 보다는, setColor의 인자로 받아서  
셀의 색을 변경해주면 더 좋지 않을까요?

sweethoneybee commented on Git-it/ViewController/SocialViewController.swift:
```diff
@@ -0,0 +1,202 @@
+        self.userCommitsSummery = try? JSONDecoder().decode(CommitsSummary.self, from: GitItApi.commitsSummary("jeong").sampleData)
+        
+        let tempFriendsCommitsSummary = try? JSONDecoder().decode([SocialCommitsSummary].self, from: GitItApi.social.sampleData)
+        self.friendsCommitsSummary = tempFriendsCommitsSummary?.sorted(by: {$0.commitsRecord.count > $1.commitsRecord.count})
```
나 + 유저를 모두 합쳐서 소팅해주세요! 
친구와 순위를 비교해보는 게 기획의도여서, 둘을 합쳐서 소팅하고 배열로 유지하는 게 좋을 것 같아요

sweethoneybee commented on Git-it/ViewController/SocialViewController.swift:
```diff
@@ -0,0 +1,202 @@
+            for friend in friends {
+                list.append(friend.username)
+            }
+            UserInfo.friendList = list
```
`UserInfo.friendList`를 갱신하는 이유가 따로 있나요? 
이유가 없다면 굳이 갱신할 필요 없을 것 같아요

parkyounbae replied:
아 이부분도 테스트하려면 일단 mock데이터에서 리스트를 받아와야해서 일단은 남겨뒀는데 수정하겠습니다~

sweethoneybee commented on Git-it/ViewController/SocialViewController.swift:
```diff
@@ -0,0 +1,202 @@
+            preconditionFailure("fail to load cell")
+        }
+        
+        switch indexPath.row {
```
dataSource에 자기 자신 결과 + 친구 결과까지 합쳐서 소팅해두면(위 리뷰 코멘트에 언급되어 있음)
switch로 나눌 거 없이 바로바로 접근 가능할 듯해요. 그리고 username도 dataSource에 같이 있으니
굳이 다시 UserInfo.friendList를 읽을 필요가 없을 듯해요

parkyounbae replied:
수정했습니다~

sweethoneybee commented on Git-it/ViewController/SocialViewController.swift:
```diff
@@ -0,0 +1,202 @@
+            cell.indexOfFriend = 1
+            cell.userCommitRecords = userCommitsSummery?.commitsRecord
+            
+            cell.addUserNameLabel()
```
cell의 UI초기화와 관련된 메소드는 여기서 호출하기보다는, 
SocialViewFriendsTableViewCell의 초기화메소드에 담아보면 어떨까요?
의미에도 더 부합하고, 여기 코드가 더 깔끔해질 듯 해요

parkyounbae replied:
수정했습니다

sweethoneybee commented on Git-it/ViewController/SocialViewController.swift:
```diff
@@ -0,0 +1,202 @@
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
```
아직 구현하지 않은 델리게이트 메소드는 삭제해두는 게 좋을 것 같아요!

sweethoneybee reviewed (changes requested):
(오 뭔가 이렇게 코드에 코멘트를 바로바로 다는 게 더 좋은 피드백이 될 수 있을 것 같네요)

구현하느라 고생하셨어요! 결과물이 잘 나왔네요. 코드 읽어보면서 궁금했던 부분이나 
더 고민해보면 좋을 것 같은 부분 코멘트 달아놨습니다.  

목업데이터 보면서 `collectionView(_:cellForItemAt:)` 메소드 구현 고민하셨을 거 생각하니
너무 죄송스럽네요 ㅠㅠ 미리 알려드렸어야 했는데.

꾸준히 개발해주셔서 감사합니다 제가 해드릴 수 있는 건 응원과 감사, 리뷰뿐.. 고맙습니다 ^^b

parkyounbae committed:
refactor: 전반적인 코드 구조 변경

SocialViewController.swift
기존에 친구와 유저의 프로퍼티가 나누어져 있던것을 SocialCommitsSummary 배열 하나로 통일
유저 정보를 받아와 SocialCommitsSummary 구조로 바꾼 뒤 배열에 추가
친구 정보를 받아와 배열에 추가한 뒤 sort함
친구/유저 로 나누어져 있던것을 통합함에 따라 셀 생성시 스위치문 사용 안함
셀 ui관련 함수를 셀 파일 안으로 옮김
완성되지 않은 삭제 관련 메소드 삭제

SocialViewFriendsTableViewCell.swift
init 관련 함수들을 맨 앞으로 옮김
dateFormatter 관련 변수를 이동
서버에서 데이터를 받아올 때 커밋을 하지 않은 날에 관련한 정보도 받아오기 때문에 셀을 생성할 때 commitrecord를 순회하면서 탐색할 필요가 없어짐
셀의 갯수만큼 데이터가 오므로 해당 셀에 해당하는 record 정보를 불러옴

GrassCollectionViewCell.swift
commitLevel프로퍼티를 없애고 setColor에서 받아오도록 수정

Resolve: GITIT-52

parkyounbae committed:
refactor: OperationQueue대신에 DispatchQueue 를 사용

간단한 작업에는 더 유리한 DispatchQueue 사용

Resolve: GITIT-52

parkyounbae committed:
fix: 친구 수 5명으로 제한

4명이던거 5명으로 수정

Resolve: GITIT-52

parkyounbae committed:
Merge remote-tracking branch 'refs/remotes/origin/main'
